File: src/main/java/no/nav/veilarbaktivitet/aktivitetskort/util/JsonUtil.java
package no.nav.veilarbaktivitet.aktivitetskort.util;

import com.fasterxml.jackson.core.JsonFactory;
import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.core.JsonToken;
import lombok.extern.slf4j.Slf4j;

import java.io.IOException;
@Slf4j
public class JsonUtil {

    private JsonUtil() {
    }

    public static String extractStringPropertyFromJson(String propertyName, String json) throws IOException {
        String result = null;
        JsonFactory jsonFactory = new JsonFactory();
        try (JsonParser parser = jsonFactory.createParser(json)) {
            parser.nextToken();
            while (parser.nextToken() != JsonToken.END_OBJECT) {
                String fieldName = parser.getCurrentName();
                if (propertyName.equals(fieldName) && parser.nextToken() == JsonToken.VALUE_STRING) {
                    result = parser.getValueAsString();
                    break;
                }
            }
        }
        return result;
    }
}

File: src/main/kotlin/no/nav/veilarbaktivitet/aktivitetskort/util/JsonUtil.kt
package no.nav.veilarbaktivitet.aktivitetskort.util

import tools.jackson.core.JsonToken
import tools.jackson.core.ObjectReadContext
import tools.jackson.core.json.JsonFactory

object JsonUtil {
    private val jsonFactory: JsonFactory = JsonFactory.builder().build()

    @JvmStatic
    fun extractStringPropertyFromJson(propertyName: String, json: String): String? {
        var result: String? = null
        val parser = jsonFactory.createParser(ObjectReadContext.empty(), json)

        fun isEnd(token: JsonToken, level: Int): Boolean {
            return token == JsonToken.END_OBJECT && level == 0
        }

        parser.use {
            var level = 0
            do {
                val nextToken = it.nextToken()
                when (nextToken) {
                    JsonToken.END_OBJECT -> { level -= 1 }
                    JsonToken.START_OBJECT -> { level += 1 }
                    JsonToken.VALUE_STRING -> {
                        val fieldName = it.currentName()
                        if (propertyName == fieldName) {
                            result = it.valueAsString
                            break
                        }
                    }
                    else -> continue
                }

            } while (!isEnd(nextToken, level))
        }
        return result
    }
}
